File: src/core/validators/Association/SecondDomainEntityPropertyMustMatchDomainOrAbstractEntity.cs
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.Association
{
    public class SecondDomainEntityPropertyMustMatchDomainOrAbstractEntity : ValidationRuleBase<MetaEdGrammar.SecondDomainEntityContext>
    {
        private readonly ISymbolTable _symbolTable;

        public SecondDomainEntityPropertyMustMatchDomainOrAbstractEntity(ISymbolTable symbolTable)
        {
            _symbolTable = symbolTable;
        }

        public override bool IsValid(MetaEdGrammar.SecondDomainEntityContext context)
        {
            var identifierToMatch = context.IdText();

            return _symbolTable.IdentifierExists(SymbolTableEntityType.DomainEntityEntityType(), identifierToMatch) ||
                _symbolTable.IdentifierExists(SymbolTableEntityType.AbstractEntityEntityType(), identifierToMatch) ||
                _symbolTable.IdentifierExists(SymbolTableEntityType.DomainEntitySubclassEntityType(), identifierToMatch);
        }

        public override string GetFailureMessage(MetaEdGrammar.SecondDomainEntityContext context)
        {
            return string.Format("Domain Entity property '{0}' does not match any declared domain or abstract entity.", context.IdText());
        }
    }
}

File: src/core/validators/Association/SecondDomainEntityPropertyMustMatchDomainOrAbstractEntity.ts
import { ValidationRuleBase } from "../ValidationRuleBase";
import {ISymbolTable} from '../SymbolTable'
export class SecondDomainEntityPropertyMustMatchDomainOrAbstractEntity extends ValidationRuleBase<MetaEdGrammar.SecondDomainEntityContext>
{
    private symbolTable: ISymbolTable;
    constructor(symbolTable: ISymbolTable) {
        super();
        this.symbolTable = symbolTable;
    }
    public isValid(context: MetaEdGrammar.SecondDomainEntityContext): boolean {
        let identifierToMatch = context.IdText();
        return this.symbolTable.identifierExists(SymbolTableEntityType.DomainEntityEntityType(), identifierToMatch) || this.symbolTable.identifierExists(SymbolTableEntityType.AbstractEntityEntityType(), identifierToMatch) || this.symbolTable.identifierExists(SymbolTableEntityType.DomainEntitySubclassEntityType(), identifierToMatch);
    }
    public getFailureMessage(context: MetaEdGrammar.SecondDomainEntityContext): string {
        return `Domain Entity property '${context.IdText()}' does not match any declared domain or abstract entity.`;
    }
}
